src/dataflow/src/sink/tail.rs

```diff
@@ -159,51 +138,36 @@ fn tail<G>(
                 }
             }
 
-            // Sort results by time and convert to Vec<Row>. We use stable sort here because
-            // it will produce deterministic results since the cursor will always produce
-            // rows in the same order.
-            results.sort_by_key(|(time, _)| *time);
-            let mut results: Vec<Row> = results.into_iter().map(|(_, row)| row).collect();
-
-            if let Some(batch) = batches.last() {
-                let progress_row = update_progress(
-                    &mut input_frontier,
-                    batch.desc.upper().borrow(),
-                    &mut packer,
-                    connector.object_columns + 1,
-                );
-                if connector.emit_progress && tail_protocol.is_some() {
-                    if let Some(progress_row) = progress_row {
-                        results.push(progress_row);
-                    }
-                }
-            }
+            // Sort results by time. We use stable sort here because it will produce deterministic
+            // results since the cursor will always producerows in the same order.
+            results.sort_by_key(|(time, _, _)| *time);
 
             // Emit data if configured, otherwise it is an error to have data to send.
             if let Some(tail_protocol) = &mut tail_protocol {
-                tail_protocol.send(results);
+                tail_protocol.send_rows(results);
             } else {
                 assert!(
                     results.is_empty(),
                     "Observed data at inactive TAIL instance"
                 );
             }
+
+            if let Some(batch) = batches.last() {
+                let progress = update_progress(&mut input_frontier, batch.desc.upper().borrow());
+                if connector.emit_progress {
+                    if let (Some(tail_protocol), Some(progress)) = (&mut tail_protocol, progress) {
+                        tail_protocol.send_progress(progress);
+                    }
+                }
+            }
```

Comment:
This logic seems similar to what was up above, but I'm a bit weirded out by it. It seems like it might only trigger on batches, which may not appear if there are empty regions of time. If you happen to know this can't happen, 1. great! and 2. we should leave some hints about why not. Otherwise, we should probably shake out whether this advances when there are no batches (due to no updates).

Reply:
I didn't look into this, but it seems this indeed implies that frontier updates are only communicated when batches are produced. The `sink` operator doesn't set `notify`, so it won't know about frontier updates.